src/Components/benchmarkapps/Wasm.Performance/TestApp/Shared/FastGrid/Grid.razor

```diff
@@ -0,0 +1,48 @@
+﻿@typeparam TRowData
+
+<CascadingValue IsFixed="true" Value="this">@ChildContent</CascadingValue>
+
+<table @attributes="@Attributes">
+    <thead>
+        <tr>
+            @foreach (var col in columns)
+            {
+                col.RenderHeader(__builder);
```

Comment:
Is this because you can't do @col.RenderHeader? without turning that into a RenderFragment first?

Reply:
Yes, you can do it with a cast `@((RenderFragment)col.RenderHeader)` but that brings extra overhead - I think (but haven't verified) that it will allocate an extra delegate, plus it will certainly go down some extra code paths in rendering. Now you raise this, I think I actually need to add `OpenRegion`/`CloseRegion` calls around this to make it fully normal as far as diffing is concerned. Will do so.

BTW the perf impacts of things like a single extra allocation, or even the difference between virtual method calls and non-virtual method calls, are really substantial on WebAssembly for high-traffic code. The most frequently-hit bits of the rendering logic that I've been profiling run about 26000 times during rendering of FastGrid. For example, switching a single method in there from non-virtual to virtual adds about 40ms to the overall render time (i.e., when invoked 26k times).

Reply:
> BTW the perf impacts of things like a single extra allocation, or even the difference between virtual method calls and non-virtual method calls, are really substantial on WebAssembly for high-traffic code. The most frequently-hit bits of the rendering logic that I've been profiling run about 26000 times during rendering of FastGrid. For example, switching a single method in there from non-virtual to virtual adds about 40ms to the overall render time (i.e., when invoked 26k times).

Can we push some of these things downstream to the runtime folks? I expect people to be using interfaces and things like that when building these types of components.

We can have guidance around it to make sure people can write the most performant code they can, but they'll have to find it. It would be best if we can make some of these things less costly.

Reply:
> From my investigations so far I already know of things we could do in the framework to speed this up. For example, two optimizations I have in mind together reduce the FastGrid times to 157ms (from blank) and 123ms (switch pages), though it remains to be seen whether we end up doing better or worse than this, as many factors may influence the end result.

Do you have a perf-goal in mind here? <100 ms would be ideal here based on [RAIL](https://web.dev/rail/#focus-on-the-user)

Reply:
> Can we push some of these things downstream to the runtime folks?

Yes, we have some ongoing conversations with the runtime people (Vlad and Zoltan mainly) about efforts to close the gap between the Mono interpreter (without jit) and CoreCLR (with jit). It's unlikely there are any bits of trivial low-hanging fruit that will give us massive gains, since there have been many optimization passes over the years. However Vlad thinks in the long term we might end up getting multiple times faster. I don't want to over-promise on his behalf though. We should expect improvements, but they will be hard-won over the long-term.

> Do you have a perf-goal in mind here? <100 ms would be ideal here based on RAIL

Telerik asked us to go for 200ms, but there really isn't any particular end state at which we can say it's done. For example, the perf gap between a decent laptop and a low-end phone might be 10x, so it depends which user's perception you most care about. I don't expect we'll reach 200ms on a very low-end phone for a 20x200 grid, as that would imply something like 20ms on desktop, and I doubt the browser can update the DOM that fast even if the .NET code took zero time. The one way we should aim to hit that target for low-end phones is all the optimizations together, plus UI virtualization.